core/libp2p/network/impl/router_impl.cpp

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "libp2p/network/impl/router_impl.hpp"
+
+OUTCOME_CPP_DEFINE_CATEGORY(libp2p::network, RouterImpl::Error, e) {
+  using E = libp2p::network::RouterImpl::Error;
+  switch (e) {
+    case E::NO_HANDLER_FOUND:
+      return "no handler was found for a given protocol";
+  }
+  return "unknown error";
+}
+
+namespace libp2p::network {
+  void RouterImpl::setProtocolHandler(const peer::Protocol &protocol,
+                                      const ProtoHandler &handler) {
+    setProtocolHandler(protocol, handler,
+                       [&protocol](const auto &p) { return protocol == p; });
```

Comment:
it is better to copy protocol inside, because it is unknown what is the lifetime of this object